### PR TITLE
Create flow between the services.

### PR DESCRIPTION
**Pull request by acceleanu**
Provide two endpoints:
- /api/payment to allow the creation of a record in the database for an amount
- /frontend/payment one to allow for the lookup of the previous amount by its payId

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -9,4 +9,4 @@ WORKDIR /app
 ADD target/*.yaml /app/
```
Comment (acceleanu) [inline Dockerfile:9]: retest this please

**Comment by acceleanu** [comment]
retest this please

File: src/main/java/uk/gov/pay/connector/util/DbConnectionChecker.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        try (Connection connection = DriverManager.getConnection(dbUrl, props)) {
+            return true;
+        } catch (Exception except) {
+            return false;
+        }
+    }
+
+    private void sleep(int seconds) {
+        try {
+            Thread.sleep(seconds);
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/util/DbConnectionChecker.java]: `Thread.sleep(...)` uses milliseconds as unit for the parameter `seconds` here, is this intentional?

File: src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            config("database.url", postgres.getConnectionUrl()),
+            config("database.user", postgres.getUsername()),
+            config("database.password", postgres.getPassword()));
+
+    @Rule
+    public RuleChain rules = RuleChain.outerRule(postgres).around(app);
+
+    @Test
+    public void makePaymentAndRetrieveAmount() throws Exception {
+        app.getApplication().run("db", "migrate", configFilePath);
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java]: Please also add the line
`connectorApp.run("db", "drop-all", "--confirm-delete-everything", configFilePath);`
similar to PaymentDaoITest.

File: src/main/java/uk/gov/pay/connector/util/DbConnectionChecker.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        Stopwatch timer = Stopwatch.createStarted();
+        boolean succeeded = false;
+        while (!succeeded && timer.elapsed(TimeUnit.SECONDS) < 10) {
+            sleep(10);
+            succeeded = checkPostgresConnection();
+        }
+        if (!succeeded) {
+            throw new RuntimeException("Postgres did not start in 10 seconds.");
+        }
+        System.out.println("Postgres docker container started in " + timer.elapsed(TimeUnit.MILLISECONDS));
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/util/DbConnectionChecker.java]: We should use a logging framework here

File: src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        int expectedAmount = 2113;
+        String payId = given().port(app.getLocalPort())
+                .contentType("application/json")
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java]: The string parameter could be replaced with `ContentType.JSON` (from the restassured library)

File: src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                .body(String.format("{\"amount\":%d}", expectedAmount))
+                .post("/api/payment")
+                .then()
+                .statusCode(200)
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java]: We should also assert the returned content type after this line with:
`.contentType(ContentType.JSON)`

File: src/main/java/uk/gov/pay/connector/resources/PaymentRequest.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
+import static java.lang.String.format;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+@Path("/api/payment")
+public class PaymentRequest {
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/PaymentRequest.java]: Class name should reflect purpose (similar to PaymentDao) - so this class could be named ChargeResource.

File: src/main/java/uk/gov/pay/connector/resources/PaymentInfo.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
+import static java.lang.String.format;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.Response.ok;
+
+@Path("/frontend/payment")
+public class PaymentInfo {
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/PaymentInfo.java]: This class should be merged with PaymentRequest (PaymentResource) - both classes handle the same resource (Charge) and should support the get and post request on the same resource URL `/v{version}/charge`
See https://docs.google.com/document/d/1OvWpE_UD91wljj7iL4oeXYLEVK_ROBLPtuJxVGCFPIg/edit
(Section Payments Store)

File: src/main/java/uk/gov/pay/connector/resources/PaymentRequest.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
+    public Response createNewPayment(JsonNode node) {
+        long amount = node.get("amount").asLong();
+
+        long payId = paymentDao.insertAmountAndReturnNewId(amount);
+
+        String response = format("{\"pay_id\":\"%s\"}", payId);
+
+        return Response.ok(response).build();
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/PaymentRequest.java]: Response status should be `201 Created` - see https://github.com/paypal/api-standards/blob/master/api-style-guide.md#create-new-resource

File: src/main/java/uk/gov/pay/connector/resources/PaymentRequest.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
+import static java.lang.String.format;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+@Path("/api/payment")
```
Comment (mikemey) [inline src/main/java/uk/gov/pay/connector/resources/PaymentRequest.java]: version number missing - according to https://github.com/paypal/api-standards/blob/master/api-style-guide.md#version - this endpoint might be named:
`/v1/charge`

**Comment by ijmad** [comment]
The language here needs to be 'charge' not 'payment'. Please change :-1:

**Comment by acceleanu** [comment]
run

File: src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                .extract()
+                .path("pay_id");
+
+        int amount = given().port(app.getLocalPort())
+                .get("/frontend/payment/" + payId)
+                .then()
+                .statusCode(200)
```
Comment (mikemey) [inline src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java]: Please assert the returned content type after this line with:
`.contentType(ContentType.JSON)`

File: src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java (hunk 3 above)
Comment (SiliconMeeple) [inline src/test/java/uk/gov/pay/connector/it/PaymentRequestTest.java]: I've extracted the common stuff out into a single rule. See what you think. :)

File: src/main/java/uk/gov/pay/connector/resources/PaymentInfo.java (hunk 8 above)
Comment (SiliconMeeple) [inline src/main/java/uk/gov/pay/connector/resources/PaymentInfo.java]: We're segregating them by consumer at the moment. It's possible that they'll merge, but if there's functionality that turns out not to be needed for the frontend / publicapi, we'd like to try not exposing it.

**Comment by SiliconMeeple** [comment]
I think that's all the changes!

**Comment by mikemey** [comment]
:+1:

**Pull request merged**